xdsl/irdl/constraints.py

```diff
@@ -55,6 +56,15 @@ def update(self, other: ConstraintContext):
 
 _AttributeCovT = TypeVar("_AttributeCovT", bound=Attribute, covariant=True)
 
+ResolveType: TypeAlias = Attribute | Sequence[Attribute]
+
+_T = TypeVar("_T")
+
+
+class Resolver(Generic[_T], abc.ABC):
+    @abstractmethod
+    def resolve(self, a: _T) -> ResolveType: ...
```

Comment:
This would benefit from a lot more doc commenting, seems like the perfect place for a long-form explanation of why this is useful and how the whole mechanism works. I'm hoping that a plain English explanation will help pick the right names for the functions and function arguments involved